[#2198 Jakarta support

azzazzel opened the pull request:
This is an attempt to automatically generate alternative modules using the new `jakarta` packages instead of `javax` ones
The approach uses [Eclipse Transformer](https://github.com/eclipse/transformer), which transforms Java binaries and can correctly map packages, type names, and related resource names.

There are four modules affected: `axon-messaging`, `axon-modelling`, `axon-eventsourcing` and `axon-configuration`. For each, there is now a respective module with the `-jakarta` suffix. 

The following applies for all of the new modules:
- **Dependencies**
   -  at module creation time, the original dependencies were manually copied from the respective POM to the new one. Thus those need to be manually synchronized if the original ones change in the future.
   - explicit `javax` dependencies were replaced by respective `jakarta` ones
   - original dependencies that transitively depended on `javax` were either updated to newer (`jakarta` compatible) versions or replaced with respective alternatives
- **Source code**
   - there is no source code in the new modules
   - compiled classes (both `main/java` and `test/java`) are copied from the respective original module with `maven-dependency-plugin`. For this to work, the original modules need to generate `test-jars`
   - copied classes are transformed in place by the `transformer-maven-plugin` 
   - transformed tests are executed by the `maven-surefire-plugin`, which allows testing the code after transformation
- **Tests**
   - the vast majority of tests pass after those changes
   - the failing tests are currently excluded in the POM. 
   - some tests rely on Spring and need to be changed as Spring does not yet support  `jakarta` packages
   - I tested the new modules with a demo I use for presentations, and it worked fine (the demo does not use JPA, though)


Considering the above, this PR is a work in progress. There are more things to do, such as:
- **investigate test failures and provide fixes**
- introduce respective integration tests
- test the new modules with a fairly complex sample app that is not based on Spring (I couldn't find one)

This pull request will resolve #2198.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/AxonFramework/AxonFramework?pullRequest=2301) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/AxonFramework/AxonFramework?pullRequest=2301) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/AxonFramework/AxonFramework?pullRequest=2301) it.</sub>

azzazzel commented:
Interesting! It finds a "bug" in `axon-integrationtests-jakarta` but not in `axon-integrationtests` module, which uses the same code. I guess SonarCloud is configured to ignore it in the original module. Anyways, I updated the dependency to `ojdbc8`, which is in Maven Central. Tests pass locally in both modules, and we can get rid of the local jar.

azzazzel commented:
Hi @smcvb 

I think this is as far as I can go with this PR. I'll mark it as ready for review now. I believe it'll be ready to merge once we rewrite the tests to not use Spring.

## Summary of the changes

- [x] new `axon-messaging-jakarta`, `axon-modelling-jakarta`, `axon-eventsourcing-jakarta`, `axon-configuration-jakarta`, and `axon-integrationtests-jakarta` modules.
- [x] made `AbstractResourceInjector` recognize `jakarta.inject.Inject`
- [x] changed `hibernate-entitymanager` to `hibernate-core` because it has a `jakarta` counterpart 
- [x] configured `axon-messaging`, `axon-modelling`, `axon-eventsourcing`, `axon-configuration`, and `axon-integrationtests` modules, to build and publish respective `test-jar` artifacts. 
- [x] replaced `@Inject` with `@Autowired` in some Spring-based tests because Spring does not recognize `@Inject` from `jakarta` package
- [x] added the Jakarta related dependencies are now in `<dependencyManagent>`. 
- [x] created `jakarta.*.legacy.version` properties and added explicit dependencies on old versions in `axon-eventsourcing`, `axon-configuration`, `axon-integrationtests`, `axon-spring`, and `axon-spring-boot-autoconfigure` modules, because Spring (and some other) depend on old jakarta jars _(those having `javax` package)_
- [x] removed local `ojdbc7.jar` and configured `axon-integrationtests` to use `ojdbc8` from Maven
- [x] bumped the version of `hibernate-validator` to `7.0.4.Final` _(which is Jakarta namespace aware)_ in all `*-jakarta` modules

## Failing tests

Currently, Spring-based, JPA tests in `*-jakarta` modules fail with

```
Caused by: java.lang.TypeNotPresentException: Type javax.persistence.EntityManagerFactory not present
```

Some of those seem complex, testing multiple threads/transactions scenarios. I'm afraid I don't have enough expertise in the area to properly change those to not use Spring. However, @Morlack offered to do it. I'll leave it between the two of you to decide how to proceed.

For convenience, here is the list of failing tests (currently excluded in the respective POMs) :
 
#### axon-messaging-jakarta

- [ ] [org.axonframework.eventhandling.tokenstore.jpa.JpaTokenStoreTest](https://github.com/AxonFramework/AxonFramework/blob/jakarta-support/messaging/src/test/java/org/axonframework/eventhandling/tokenstore/jpa/JpaTokenStoreTest.java)

#### axon-modelling-jakarta

- [ ] [org.axonframework.modelling.saga.repository.jpa.JpaSagaStoreTest](https://github.com/AxonFramework/AxonFramework/blob/jakarta-support/modelling/src/test/java/org/axonframework/modelling/saga/repository/jpa/JpaSagaStoreTest.java)

#### axon-eventsourcing-jakarta

- [ ] [org.axonframework.eventsourcing.eventstore.jpa.JpaEventStorageEngineTest](https://github.com/AxonFramework/AxonFramework/blob/jakarta-support/eventsourcing/src/test/java/org/axonframework/eventsourcing/eventstore/jpa/JpaEventStorageEngineTest.java)

#### axon-integrationtests-jakarta

- [ ] [org.axonframework.integrationtests.eventsourcing.eventstore.benchmark.JpaStorageEngineInsertionReadOrderTest](https://github.com/AxonFramework/AxonFramework/blob/jakarta-support/integrationtests/src/test/java/org/axonframework/integrationtests/eventsourcing/eventstore/benchmark/JpaStorageEngineInsertionReadOrderTest.java)
- [ ] [org.axonframework.integrationtests.modelling.command.GenericJpaRepositoryIntegrationTest](https://github.com/AxonFramework/AxonFramework/blob/jakarta-support/integrationtests/src/test/java/org/axonframework/integrationtests/modelling/command/GenericJpaRepositoryIntegrationTest.java)

smcvb commented on config-jakarta/pom.xml:
```diff
@@ -0,0 +1,211 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Missing the [copyright notice](https://github.com/AxonFramework/AxonFramework/blob/master/copyright-template.xml) in this file.

azzazzel replied:
Added

smcvb commented on eventsourcing-jakarta/pom.xml:
```diff
@@ -0,0 +1,290 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Missing the [copyright notice](https://github.com/AxonFramework/AxonFramework/blob/master/copyright-template.xml) in this file.

azzazzel replied:
Added

smcvb commented on integrationtests-jakarta/pom.xml:
```diff
@@ -0,0 +1,293 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Missing the [copyright notice](https://github.com/AxonFramework/AxonFramework/blob/master/copyright-template.xml) in this file.

azzazzel replied:
Added

smcvb commented on integrationtests-jakarta/pom.xml:
```diff
@@ -0,0 +1,293 @@
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>            <plugin>
```
Nit: Identation is a bit off here

azzazzel replied:
Fixed

smcvb commented on integrationtests-jakarta/pom.xml:
```diff
@@ -0,0 +1,293 @@
+                        </configuration>
+                    </execution>
+                </executions>
+                <configuration>
```
Just marking it with a comment, but this needs to be removed once we've made the tests work with Jakarta.

smcvb commented on messaging-jakarta/pom.xml:
```diff
@@ -0,0 +1,332 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Missing the [copyright notice](https://github.com/AxonFramework/AxonFramework/blob/master/copyright-template.xml) in this file.

azzazzel replied:
Added

smcvb commented on messaging-jakarta/pom.xml:
```diff
@@ -0,0 +1,332 @@
+                <configuration>
+                    <trimStackTrace>false</trimStackTrace>
+                    <useFile>false</useFile>
+                    <excludes>
```
Just marking it with a comment, but this needs to be removed once we've made the tests work with Jakarta.

smcvb commented on modelling-jakarta/pom.xml:
```diff
@@ -0,0 +1,302 @@
+                    </execution>
+                </executions>
+                <configuration>
+                    <excludes>
```
Just marking it with a comment, but this needs to be removed once we've made the tests work with Jakarta.

smcvb commented on modelling-jakarta/src/main/resources/perClassConstants.properties:
```diff
@@ -0,0 +1 @@
+org/axonframework/modelling/saga/AbstractResourceInjector.class=notransform.properties
```
Curious: I expect you need to add the `AbstractResourceInjector` to be ignored for the `DEFAULT_INJECT_ANNOTATIONS` constant that contains `javax.inject.Inject`, right?

smcvb replied:
Or is that already solved by the `notransform.properties` file? :thinking:

azzazzel replied:
Correct. By default, it uses the default Jakarta rules (see `<jakartaDefaults>true</jakartaDefaults>` configured in `pluginManagement` in parent POM). For `AbstractResourceInjector`, we need a custom rule (that is what `notransform.properties` is) on top of that because the class needs to be aware of both namespaces.

smcvb replied:
Awesome, thanks for clarifying that, @azzazzel!

smcvb commented on modelling/src/main/java/org/axonframework/modelling/saga/AbstractResourceInjector.java:
```diff
@@ -42,6 +42,7 @@ public abstract class AbstractResourceInjector implements ResourceInjector {
 
     private static final String[] DEFAULT_INJECT_ANNOTATIONS = {
             "javax.inject.Inject",
+            "jakarta.inject.Inject",
```
Having a test for this would be nice.

azzazzel replied:
That could be troublesome, I guess. You'd need to add `jakarta` packages to this non-jakarta module, for no other purpose but this test.  

Also, is it really needed? I assume there is a test for the `javax` package. If so, it is automatically copied to the respective `jakarta` module, transformed, and run there. So effectively, there is a test for `jakarta.inject.Inject` as well. Just that we don't have the source code for it.

smcvb replied:
You're right, the `SimpleResourceInjectorTest` should pick this up. Good thinking :+1:

smcvb commented on pom.xml:
```diff
@@ -398,6 +426,32 @@
                 <version>${hamcrest.version}</version>
             </dependency>
+
+            <dependency>
```
Super nIt: why not pair up the Javax and Jakarta dependencies next to one another like you did with `org.glassfish`?

azzazzel replied:
No reason. It was just convenient to track what was missing while working on it. I paired the dependencies up now.

smcvb reviewed (changes requested):
Bunch of minor comments and questions. Mainly blocking for the tests that as pointed out aren't done yet, and the missing copyright notices in the new `pom.xml` files.

smcvb commented on messaging-jakarta/pom.xml:
```diff
@@ -0,0 +1,348 @@
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>            <plugin>
```
Nit: identation is slightly off here.
